src/drivers/rtmouse_dev.c

```diff
@@ -0,0 +1,923 @@
+/*
+ *
+ * rtmouse_dev.c
+ * Define device file operations and register device
+ *
+ * Copyright (C) 2024 RT Corporation <shop@rt-net.jp>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#include "rtmouse.h"
+
+static unsigned int motor_l_freq_is_positive = 1;
+static unsigned int motor_r_freq_is_positive = 1;
+
+/*
+ * --- Device Names(+%u) ---
+ * used in register_dev()
+ */
+static const char *NAME_DEV_U[ID_DEV_SIZE] = {
+    [ID_DEV_LED] = "rtled%u",
+    [ID_DEV_SWITCH] = "rtswitch%u",
+    [ID_DEV_SENSOR] = "rtlightsensor%u",
+    [ID_DEV_BUZZER] = "rtbuzzer%u",
+    [ID_DEV_MOTORRAWR] = "rtmotor_raw_r%u",
+    [ID_DEV_MOTORRAWL] = "rtmotor_raw_l%u",
+    [ID_DEV_MOTOREN] = "rtmotoren%u",
+    [ID_DEV_MOTOR] = "rtmotor%u"};
+
+/*
+ * i2c_counter_set - set value to I2C pulse counter
+ * called by rtcnt_write()
+ */
+static int i2c_counter_set(struct rtcnt_device_info *dev_info, int setval)
+{
+	int ret = 0;
+	int lsb = 0, msb = 0;
+	struct i2c_client *client = dev_info->client;
+
+	// printk(KERN_INFO "set 0x%x = 0x%x\n", client->addr, setval);
+	msb = (setval >> 8) & 0xFF;
+	lsb = setval & 0xFF;
+	mutex_lock(&dev_info->lock);
+	// printk(KERN_INFO "set 0x%x msb = 0x%x\n", client->addr, msb);
+	ret = i2c_smbus_write_byte_data(client, 0x10, msb);
+	if (ret < 0) {
+		printk(KERN_ERR
+		       "%s: Failed writing to i2c counter device, addr=0x%x\n",
+		       __func__, client->addr);
+		return -ENODEV;
+	}
+	// printk(KERN_INFO "set 0x%x lsb = 0x%x\n", client->addr, lsb);
+	ret = i2c_smbus_write_byte_data(client, 0x11, lsb);
+	if (ret < 0) {
+		printk(KERN_ERR
+		       "%s: Failed writing to i2c counter device, addr=0x%x\n",
+		       __func__, client->addr);
+		return -ENODEV;
+	}
+	mutex_unlock(&dev_info->lock);
+	return ret;
+}
+
+/*
+ * i2c_counter_read - get value from I2C pulse counter
+ * called by rtcnt_read()
+ */
+static int i2c_counter_read(struct rtcnt_device_info *dev_info, int *ret)
+{
+	int lsb = 0, msb = 0;
+	// printk(KERN_INFO "read 0x%x\n", client->addr);
+	struct i2c_client *client = dev_info->client;
+	mutex_lock(&dev_info->lock);
+
+	lsb = i2c_smbus_read_byte_data(client, CNT_ADDR_LSB);
+	if (lsb < 0) {
+		printk(
+		    KERN_ERR
+		    "%s: Failed reading from i2c counter device, addr=0x%x\n",
+		    __func__, client->addr);
+		return -ENODEV;
```

Comment:
エラーハンドリングをした際に`mutex_unlock(&dev_info->lock);`が呼ばれず、リソースリークを起こす可能性があるため、returnの前に追加します。

```suggestion
　　　　mutex_unlock(&dev_info->lock);
		return -ENODEV;
```